### PR TITLE
Feature/security analysis

### PR DESCRIPTION
**Pull request by armic-artelys**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
- Added internal classes to perform integration tests on SA.
- Added SA testcase file in test resources.
- Follows the previous PR : https://github.com/powsybl/powsybl-integration-test/pull/10 that adds several test cases for Loadflow.

**What is the current behavior?** *(You can also link to an open issue here)*
Only Loadflow tests can be performed.

**What is the new behavior (if this is a feature change)?**
SA Tests can now be added using the JSON configuration file `src/test/resources/saTestPlan.json`. 
SA Tests are run by the SAIntegrationTest class.

### CODE REVIEW

**Comment by armic-artelys** [comment]
@obrix : PR is now ready to be reviewed.

File: src/test/resources/saTestPlan.json [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{
+  "version" : "1.0",
+  "testCases" : [
+    {
+      "id" : "6-noeuds-dummy",
+      "inputNetwork" : "networks/reseau_6noeuds.xiidm",
+      "inputParameters" : "parameters/SAParametersBasic.json",
+      "inputContingencies": "contingencies/6noeuds-dummy.json",
+      "expectedNetwork" : "sareferences/6noeuds_dummy.xiidm",
```
Comment (obrix) [inline src/test/resources/saTestPlan.json:9]: I think you lack a state monitor attribute to actually monitor network element at the pre / post contingency state (see 6noeuds_dummy.json networkResults are empty).

File: src/main/java/com/powsybl/integrationtest/security/model/SAComputationParameters.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ * Security Analysis computation parameters. Contains the network and the {@link SecurityAnalysisParameters} to use to
+ * run the computation.
+ *
+ * @author Arthur Michaut <arthur.michaut at artelys.com>
+ */
+public class SAComputationParameters implements ComputationParameters {
+    private final SecurityAnalysisParameters saParameters;
+    private final Network network;
+    private final List<Contingency> contingencies;
+
```
Comment (obrix) [inline src/main/java/com/powsybl/integrationtest/security/model/SAComputationParameters.java:26]: List of StateMonitor to be added here probably.

File: src/main/java/com/powsybl/integrationtest/creation/loadflow/LoadflowTestcaseCreator.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+import java.nio.file.Path;
+import java.util.Properties;
+
+/**
+ * A component that creates loadflow test cases reference files.
+ * Use the main method, linking to a parameters file to output your own reference files (network and results).
+ *
+ * @author Arthur Michaut <arthur.michaut at artelys.com>
+ */
+public class LoadflowTestcaseCreator {
```
Comment (obrix) [inline src/main/java/com/powsybl/integrationtest/creation/loadflow/LoadflowTestcaseCreator.java:33]: I think it should be LoadFlowTestCaseCreator (see LoadFlow API in the core and TestCase here)

File: src/main/java/com/powsybl/integrationtest/creation/loadflow/LoadflowTestcaseCreatorParameters.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.List;
+
+/**
+ * Parameters used by the Loadflow test cases creator. Made to be loaded from a JSON file
+ *
+ * @author Arthur Michaut <arthur.michaut at artelys.com>
+ */
+class LoadflowTestcaseCreatorParameters {
```
Comment (obrix) [inline src/main/java/com/powsybl/integrationtest/creation/loadflow/LoadflowTestcaseCreatorParameters.java:22]: LoadFlowTestCaseCreatorParameters

File: src/main/java/com/powsybl/integrationtest/creation/security/SATestcaseCreator.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+import java.nio.file.Path;
+import java.util.Properties;
+
+/**
+ * A component that creates SA test cases reference files.
+ * Use the main method, linking to a parameters file to output your own reference files (network and results).
+ *
+ * @author Arthur Michaut <arthur.michaut at artelys.com>
+ */
+public class SATestcaseCreator {
```
Comment (obrix) [inline src/main/java/com/powsybl/integrationtest/creation/security/SATestcaseCreator.java:34]: SecurityAnalysisTestCaseCreator

File: src/main/java/com/powsybl/integrationtest/creation/security/SATestcaseCreatorParameters.java [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.List;
+
+/**
+ * Parameters used by the SA test cases creator. Made to be loaded from a JSON file
+ *
+ * @author Arthur Michaut <arthur.michaut at artelys.com>
+ */
+public class SATestcaseCreatorParameters {
```
Comment (obrix) [inline src/main/java/com/powsybl/integrationtest/creation/security/SATestcaseCreatorParameters.java:22]: SecurityAnalysisTestCaseCreatorParameters

File: src/main/java/com/powsybl/integrationtest/loadflow/jsonconfig/LoadflowTestCaseJsonDeserializer.java [hunk 7]
```diff
@@ -20,7 +25,45 @@ protected LoadflowTestCaseJsonDeserializer() {
     }
 
     @Override
-    public LoadflowTestCaseJson deserialize(JsonParser jsonParser, DeserializationContext deserializationContext) {
-        return LoadflowTestCaseJson.parseJson(jsonParser);
+    public LoadflowTestCaseJson deserialize(JsonParser parser, DeserializationContext context) throws IOException {
```
Comment (obrix) [inline src/main/java/com/powsybl/integrationtest/loadflow/jsonconfig/LoadflowTestCaseJsonDeserializer.java:28]: LoadFlowTestCaseJson

File: src/main/java/com/powsybl/integrationtest/loadflow/model/LoadflowComputationRunner.java [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+import com.powsybl.loadflow.LoadFlow;
+import com.powsybl.loadflow.LoadFlowResult;
+
+/**
+ * A {@link ComputationRunner} for Loadflow calculations. Provide it with a {@link LoadflowComputationParameters}
+ * and it will return a {@link LoadflowComputationResults}.
+ *
+ * @author Arthur Michaut <arthur.michaut at artelys.com>
+ */
+public class LoadflowComputationRunner implements ComputationRunner<LoadflowComputationParameters, LoadflowComputationResults> {
```
Comment (obrix) [inline src/main/java/com/powsybl/integrationtest/loadflow/model/LoadflowComputationRunner.java:19]: LoadFlowComputationRunner

File: src/main/java/com/powsybl/integrationtest/security/jsonconfig/SATestCaseJson.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+import javax.annotation.Nonnull;
+import java.util.Objects;
+
+/**
+ * DTO for the JSON representation of an SA Test Case.
+ *
+ * @author Arthur Michaut <arthur.michaut at artelys.com>
+ */
+public class SATestCaseJson {
```
Comment (obrix) [inline src/main/java/com/powsybl/integrationtest/security/jsonconfig/SATestCaseJson.java:17]: SecurityAnalysisTestCaseJson

File: src/main/java/com/powsybl/integrationtest/security/jsonconfig/SATestCaseJsonDeserializer.java [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * Deserializer for JSON representations of {@link SATestCaseJson}
+ *
+ * @author Arthur Michaut <arthur.michaut at artelys.com>
+ */
+public class SATestCaseJsonDeserializer extends StdDeserializer<SATestCaseJson> {
```
Comment (obrix) [inline src/main/java/com/powsybl/integrationtest/security/jsonconfig/SATestCaseJsonDeserializer.java:23]: SecurityAnalysisTestCaseJsonDeserializer

File: src/main/java/com/powsybl/integrationtest/security/jsonconfig/SATestPlanJson.java [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+package com.powsybl.integrationtest.security.jsonconfig;
+
+import java.util.List;
+
+/**
+ * A test plan that contains SA test cases.
+ *
+ * @author Arthur Michaut <arthur.michaut at artelys.com>
+ */
+public class SATestPlanJson {
```
Comment (obrix) [inline src/main/java/com/powsybl/integrationtest/security/jsonconfig/SATestPlanJson.java:16]: SecurityAnalysisTestPlanJson

File: src/main/java/com/powsybl/integrationtest/security/jsonconfig/SATestPlanJsonDeserializer.java [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Deserializer for JSON representations of {@link SATestPlanJson}
+ *
+ * @author Arthur Michaut <arthur.michaut at artelys.com>
+ */
+public class SATestPlanJsonDeserializer extends StdDeserializer<SATestPlanJson> {
```
Comment (obrix) [inline src/main/java/com/powsybl/integrationtest/security/jsonconfig/SATestPlanJsonDeserializer.java:24]: SecurityAnalysisTestPlanJsonDeserializer

File: src/main/java/com/powsybl/integrationtest/security/jsonconfig/SATestPlanReader.java [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * JSON Reader for Security Analysis Test Plans.
+ *
+ * @author Arthur Michaut <arthur.michaut at artelys.com>
+ */
+public class SATestPlanReader implements TestPlanReader<SAComputationParameters, SAComputationResults, SATestCase> {
```
Comment (obrix) [inline src/main/java/com/powsybl/integrationtest/security/jsonconfig/SATestPlanReader.java:40]: SecurityAnalysisTestPlanReader

File: src/main/java/com/powsybl/integrationtest/security/model/SAComputationParameters.java [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+import java.util.List;
+
+/**
+ * Security Analysis computation parameters. Contains the network and the {@link SecurityAnalysisParameters} to use to
+ * run the computation.
+ *
+ * @author Arthur Michaut <arthur.michaut at artelys.com>
+ */
+public class SAComputationParameters implements ComputationParameters {
```
Comment (obrix) [inline src/main/java/com/powsybl/integrationtest/security/model/SAComputationParameters.java:22]: SecurityAnalysisComputationParameters

File: src/main/java/com/powsybl/integrationtest/security/model/SAComputationResults.java [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+import com.powsybl.integrationtest.model.ComputationResults;
+import com.powsybl.security.SecurityAnalysisResult;
+
+/**
+ * Security Analysis computation results. Contains the network (post-contingencies) and the corresponding
+ * {@link SecurityAnalysisResult}.
+ *
+ * @author Arthur Michaut <arthur.michaut at artelys.com>
+ */
+public class SAComputationResults implements ComputationResults {
```
Comment (obrix) [inline src/main/java/com/powsybl/integrationtest/security/model/SAComputationResults.java:19]: SecurityAnalysisComputationResults

File: src/main/java/com/powsybl/integrationtest/security/model/SAComputationRunner.java [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+import com.powsybl.integrationtest.model.ComputationRunner;
+import com.powsybl.security.SecurityAnalysis;
+import com.powsybl.security.SecurityAnalysisReport;
+
+/**
+ * A {@link ComputationRunner} that runs Security Analysis calculations.
+ *
+ * @author Arthur Michaut <arthur.michaut at artelys.com>
+ */
+public class SAComputationRunner implements ComputationRunner<SAComputationParameters, SAComputationResults> {
```
Comment (obrix) [inline src/main/java/com/powsybl/integrationtest/security/model/SAComputationRunner.java:18]: SecurityAnalysisComputationRunner

File: src/main/java/com/powsybl/integrationtest/security/model/SATestCase.java [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+
+import com.powsybl.integrationtest.model.TestCase;
+
+/**
+ * A {@link TestCase} for Security Analysis computations. Holds a {@link SAComputationParameters} and
+ * a {@link SAComputationResults}. Can be used with an {@link SATestRunner}.
+ *
+ * @author Arthur Michaut <arthur.michaut at artelys.com>
+ */
+public class SATestCase implements TestCase<SAComputationParameters, SAComputationResults> {
```
Comment (obrix) [inline src/main/java/com/powsybl/integrationtest/security/model/SATestCase.java:17]: SecurityAnalysisTestCase

File: src/main/java/com/powsybl/integrationtest/security/model/SATestRunner.java [hunk 18]
```diff
@@ -0,0 +1,256 @@
...
+
+import static com.powsybl.integrationtest.utils.CompareUtils.assertDeltaMax;
+import static com.powsybl.integrationtest.utils.CompareUtils.assertEquals;
+
+/**
+ * A {@link TestRunner} for Security Analysis
+ *
+ * @author Arthur Michaut <arthur.michaut at artelys.com>
+ */
+public class SATestRunner extends AbstractTestRunner<SAComputationParameters, SAComputationResults> {
```
Comment (obrix) [inline src/main/java/com/powsybl/integrationtest/security/model/SATestRunner.java:30]: SecurityAnalysisTestRunner

**Review by obrix — CHANGES_REQUESTED** [review]
First review : 
* StateMonitor configuration is missing from the SAComputationParameters leading to empty network results.
* To be coherent with LoadFlow API from the core I think all Loadflow* shoul actually by LoadFlow*
* Also to be coherent with the core and to avoid futur ambiguity with SensitivityAnalysis I think all SA* class should be called more explicitly SecurityAnalysis*

File: src/test/resources/saTestPlan.json (hunk 1 above)
Comment (armic-artelys) [reply src/test/resources/saTestPlan.json:9]: My bad : added them.

File: src/main/java/com/powsybl/integrationtest/security/model/SAComputationParameters.java (hunk 2 above)
Comment (armic-artelys) [reply src/main/java/com/powsybl/integrationtest/security/model/SAComputationParameters.java:26]: true

File: src/main/java/com/powsybl/integrationtest/creation/loadflow/LoadflowTestcaseCreator.java (hunk 3 above)
Comment (armic-artelys) [reply src/main/java/com/powsybl/integrationtest/creation/loadflow/LoadflowTestcaseCreator.java]: Fixed

File: src/main/java/com/powsybl/integrationtest/creation/loadflow/LoadflowTestcaseCreatorParameters.java (hunk 4 above)
Comment (armic-artelys) [reply src/main/java/com/powsybl/integrationtest/creation/loadflow/LoadflowTestcaseCreatorParameters.java]: Fixed

File: src/main/java/com/powsybl/integrationtest/creation/security/SATestcaseCreator.java (hunk 5 above)
Comment (armic-artelys) [reply src/main/java/com/powsybl/integrationtest/creation/security/SATestcaseCreator.java:36]: done

File: src/main/java/com/powsybl/integrationtest/creation/security/SATestcaseCreatorParameters.java (hunk 6 above)
Comment (armic-artelys) [reply src/main/java/com/powsybl/integrationtest/creation/security/SATestcaseCreatorParameters.java:22]: done

File: src/main/java/com/powsybl/integrationtest/loadflow/jsonconfig/LoadflowTestCaseJsonDeserializer.java (hunk 7 above)
Comment (armic-artelys) [reply src/main/java/com/powsybl/integrationtest/loadflow/jsonconfig/LoadflowTestCaseJsonDeserializer.java]: done

File: src/main/java/com/powsybl/integrationtest/security/jsonconfig/SATestCaseJson.java (hunk 9 above)
Comment (armic-artelys) [reply src/main/java/com/powsybl/integrationtest/security/jsonconfig/SATestCaseJson.java]: done

File: src/main/java/com/powsybl/integrationtest/security/jsonconfig/SATestPlanJson.java (hunk 11 above)
Comment (armic-artelys) [reply src/main/java/com/powsybl/integrationtest/security/jsonconfig/SATestPlanJson.java]: done

File: src/main/java/com/powsybl/integrationtest/security/jsonconfig/SATestPlanReader.java (hunk 13 above)
Comment (armic-artelys) [reply src/main/java/com/powsybl/integrationtest/security/jsonconfig/SATestPlanReader.java]: done

File: src/main/java/com/powsybl/integrationtest/security/model/SAComputationResults.java (hunk 15 above)
Comment (armic-artelys) [reply src/main/java/com/powsybl/integrationtest/security/model/SAComputationResults.java]: done

File: src/main/java/com/powsybl/integrationtest/security/model/SAComputationRunner.java (hunk 16 above)
Comment (armic-artelys) [reply src/main/java/com/powsybl/integrationtest/security/model/SAComputationRunner.java]: done

File: src/main/java/com/powsybl/integrationtest/security/model/SATestCase.java (hunk 17 above)
Comment (armic-artelys) [reply src/main/java/com/powsybl/integrationtest/security/model/SATestCase.java]: done

File: src/main/java/com/powsybl/integrationtest/security/model/SATestRunner.java (hunk 18 above)
Comment (armic-artelys) [reply src/main/java/com/powsybl/integrationtest/security/model/SATestRunner.java]: done

File: src/main/java/com/powsybl/integrationtest/security/jsonconfig/SATestPlanJsonDeserializer.java (hunk 12 above)
Comment (armic-artelys) [reply src/main/java/com/powsybl/integrationtest/security/jsonconfig/SATestPlanJsonDeserializer.java]: done

File: src/main/java/com/powsybl/integrationtest/security/model/SAComputationParameters.java (hunk 14 above)
Comment (armic-artelys) [reply src/main/java/com/powsybl/integrationtest/security/model/SAComputationParameters.java]: done

**Comment by armic-artelys** [comment]
I have done the changes : you can keep-on with the review.

**Review by obrix — APPROVED** [review]

**Pull request closed without merging**